File: shell_wrapper/kahe.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust wrapper around the simplified C++ API for Key Additive Homomorphic
//! Encryption.

use shell_types::{Moduli, RnsContextRef, RnsPolynomial, RnsPolynomialVec};
use single_thread_hkdf::{SeedWrapper, SingleThreadHkdfWrapper};
use status::rust_status_from_cpp;
use std::collections::HashMap;
use std::marker::PhantomData;
use std::mem::MaybeUninit;

#[derive(Debug, PartialEq, Clone)]
pub struct PackedVectorConfig {
    pub base: u64,
    pub dimension: u64,
    pub num_packed_coeffs: u64,
}

#[cxx::bridge]
mod ffi {
    /// Owned KahePublicParameters behind a unique_ptr.
    pub struct KahePublicParametersWrapper {
        pub ptr: UniquePtr<KahePublicParameters>,
    }

    pub struct BigIntVectorWrapper {
        pub ptr: UniquePtr<CxxVector<BigInteger>>,
    }

    unsafe extern "C++" {
        include!("shell_wrapper/kahe.h");
        include!("shell_wrapper/shell_types.h");

        #[namespace = "secure_aggregation"]
        type KahePublicParameters;

        #[namespace = "secure_aggregation"]
        type BigInteger;

        type FfiStatus = shell_types::ffi::FfiStatus;
        type ModuliWrapper = shell_types::ffi::ModuliWrapper;
        #[namespace = "secure_aggregation"]
        type RnsContext = shell_types::ffi::RnsContext;
        type RnsPolynomialWrapper = shell_types::ffi::RnsPolynomialWrapper;
        type RnsPolynomialVecWrapper = shell_types::ffi::RnsPolynomialVecWrapper;
        type SingleThreadHkdfWrapper = single_thread_hkdf::SingleThreadHkdfWrapper;

        pub unsafe fn CreateKahePublicParametersWrapper(
            log_n: u64,
            log_t: u64,
            qs: &[u64],
            num_public_polynomials: u64,
            seed: &[u8],
            out: *mut KahePublicParametersWrapper,
        ) -> FfiStatus;

        pub unsafe fn CreateModuliWrapperFromKaheParams(
            params: &KahePublicParametersWrapper,
        ) -> ModuliWrapper;

        pub unsafe fn GetRnsContextFromKaheParams(
            params: &KahePublicParametersWrapper,
        ) -> *const RnsContext;

        pub unsafe fn GenerateSecretKeyWrapper(
            params: &KahePublicParametersWrapper,
            prng: *mut SingleThreadHkdfWrapper,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;

        pub unsafe fn PackMessagesRaw(
            messages: &[u64],
            packing_base: u64,
            packing_dimension: u64,
            num_packed_values: u64,
            packed_values: *mut BigIntVectorWrapper,
        ) -> FfiStatus;

        pub unsafe fn UnpackMessagesRaw(
            packing_base: u64,
            packing_dimension: u64,
            num_packed_values: u64,
            packed_values: &mut BigIntVectorWrapper,
            out: &mut Vec<u64>,
        ) -> FfiStatus;

        pub unsafe fn Encrypt(
            packed_values: &BigIntVectorWrapper,
            secret_key: &RnsPolynomialWrapper,
            params: &KahePublicParametersWrapper,
            prng: *mut SingleThreadHkdfWrapper,
            out: *mut RnsPolynomialVecWrapper,
        ) -> FfiStatus;

        pub unsafe fn Decrypt(
            ciphertexts: &RnsPolynomialVecWrapper,
            secret_key: &RnsPolynomialWrapper,
            params: &KahePublicParametersWrapper,
            output_values: *mut BigIntVectorWrapper,
        ) -> FfiStatus;
    }
}
pub use ffi::KahePublicParametersWrapper;

/// Creates new public parameters for KAHE. `num_public_polynomials` is the
/// number of public "a" polynomials to generate from the public `seed`. Each
/// call to `encrypt` using the same secret key must use a different public
/// polynomial. `log_t` is the number of bits of the KAHE plaintext modulus (q1
/// from [Willow](https://eprint.iacr.org/2024/936.pdf)).
pub fn create_public_parameters(
    log_n: u64,
    log_t: u64,
    qs: &[u64],
    num_public_polynomials: usize,
    seed: &SeedWrapper,
) -> Result<KahePublicParametersWrapper, status::StatusError> {
    let mut out = MaybeUninit::<KahePublicParametersWrapper>::zeroed();
    // SAFETY: No lifetime constraints (the new `PublicParametersWrapper` does not
    // keep any reference to the seed). Only reads the `qs` buffer within a valid
    // range.
    rust_status_from_cpp(unsafe {
        ffi::CreateKahePublicParametersWrapper(
            log_n,
            log_t,
            qs,
            num_public_polynomials as u64,
            seed.as_bytes(),
            out.as_mut_ptr(),
        )
    })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Returns RNS moduli, containing pointers to the moduli in the public
/// parameters.
pub fn get_moduli<'a>(params: &'a KahePublicParametersWrapper) -> Moduli<'a> {
    // SAFETY: `moduli` contains raw pointers to the moduli in `params`, but the bindings
    // don't know that, so we add a lifetime annotation with PhantomData. After
    // that, both `params` and `moduli` live for at least `'a`.
    let moduli = unsafe { ffi::CreateModuliWrapperFromKaheParams(params) };
    Moduli { moduli, phantom: PhantomData }
}

/// Returns an RnsContextRef, containing a pointer to the RNS context in the
/// public parameters.
pub fn get_rns_context_ref<'a>(params: &'a KahePublicParametersWrapper) -> RnsContextRef<'a> {
    // SAFETY: `rns_context` contains a raw pointer to the RNS context in
    // `params`, but the bindings don't know that, so we add a lifetime annotation
    // with PhantomData. After that, both `params` and `rns_context` live for at
    // least `
    let rns_context = unsafe { ffi::GetRnsContextFromKaheParams(params) };
    RnsContextRef { rns_context, phantom: PhantomData }
}

/// Generates a BGV secret key.
pub fn generate_secret_key(
    params: &KahePublicParametersWrapper,
    prng: &mut SingleThreadHkdfWrapper,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    // SAFETY: `out` pointer is valid, no references to `params` or `prng` are kept.
    rust_status_from_cpp(unsafe { ffi::GenerateSecretKeyWrapper(params, prng, out.as_mut_ptr()) })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

pub use ffi::BigIntVectorWrapper;

/// Encrypts the vectors stored in `input_vectors` using `secret_key` and the public polynomials
/// stored in `params`. The input vectors are packed according to the given `packed_vector_configs`.
/// Returns the resulting ciphertexts.
pub fn encrypt(
    input_vectors: &HashMap<String, Vec<u64>>,
    packed_vector_configs: &HashMap<String, PackedVectorConfig>,
    secret_key: &RnsPolynomial,
    params: &KahePublicParametersWrapper,
    prng: &mut SingleThreadHkdfWrapper,
) -> Result<RnsPolynomialVec, status::StatusError> {
    let mut packed_values = MaybeUninit::<BigIntVectorWrapper>::zeroed();
    // SAFETY: No lifetime constraints (`PackMessagesRaw` may create a new vector of BigIntegers
    // wrapped by `packed_values` which does not keep any reference to the inputs).
    // `PackMessagesRaw` only appends to the C++ vector wrapped by `packed_values`,
    // allocating it in case it is NULL (in the first iteration).
    for (id, packed_vector_config) in packed_vector_configs.iter() {
        if !input_vectors.contains_key(id) {
            return Err(status::invalid_argument(format!("Input vector with id {} not found", id)));
        }
        rust_status_from_cpp(unsafe {
            ffi::PackMessagesRaw(
                &input_vectors[id],
                packed_vector_config.base,
                packed_vector_config.dimension,
                packed_vector_config.num_packed_coeffs,
                packed_values.as_mut_ptr(),
            )
        })?;
    }

    let mut out = MaybeUninit::<RnsPolynomialVec>::zeroed();
    // SAFETY: No lifetime constraints (`Encrypt` creates a new vector of polynomials wrapped by
    // `out` which does not keep any reference to the inputs). `Encrypt` reads the C++ vector
    // wrapped by `packed_values`, updates the states wrapped by `prng`, and writes into the C++
    // vector wrapped by `out`.
    rust_status_from_cpp(unsafe {
        ffi::Encrypt(&packed_values.assume_init(), secret_key, params, prng, out.as_mut_ptr())
    })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Decrypts ciphertexts that were encrypted with `secret_key` and the public polynomials stored
/// in `params`. Returns the unpacked decrypted values.
/// The decrypted values are unpacked according to the given `packed_vector_configs`.
pub fn decrypt(
    ciphertext: &RnsPolynomialVec,
    secret_key: &RnsPolynomial,
    params: &KahePublicParametersWrapper,
    packed_vector_configs: &HashMap<String, PackedVectorConfig>,
) -> Result<HashMap<String, Vec<u64>>, status::StatusError> {
    let mut packed_values = MaybeUninit::<BigIntVectorWrapper>::zeroed();
    // SAFETY: No lifetime constraints (`packed_values` does not keep any reference to the inputs).
    // `Decrypt` creates a new C++ vector wrapped by `output_values` and only modifies this buffer.
    rust_status_from_cpp(unsafe {
        ffi::Decrypt(ciphertext, secret_key, params, packed_values.as_mut_ptr())
    })?;

    let mut output_vectors = HashMap::<String, Vec<u64>>::new();
    // Assume the packed values are stored in the same order as the configs.
    for (id, packed_vector_config) in packed_vector_configs.iter() {
        let unpacked_size =
            (packed_vector_config.num_packed_coeffs * packed_vector_config.dimension) as usize;
        let mut unpacked_values = Vec::with_capacity(unpacked_size);
        /// SAFETY: No lifetime constraints (output values of `UnpackMessagesRaw` do not keep any
        /// reference to its inputs). `UnpackMessagesRaw` reads and removes a prefix of the C++
        /// vector wrapped by `packed_values`, and writes into the buffer `out`.
        rust_status_from_cpp(unsafe {
            ffi::UnpackMessagesRaw(
                packed_vector_config.base,
                packed_vector_config.dimension,
                packed_vector_config.num_packed_coeffs,
                packed_values.assume_init_mut(),
                &mut unpacked_values,
            )
        })?;
        output_vectors.insert(id.clone(), unpacked_values);
    }

    Ok(output_vectors)
}

File: shell_wrapper/kahe.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/kahe.h"

#include <sys/types.h>

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/types/span.h"
#include "include/cxx.h"
#include "shell_encryption/rns/coefficient_encoder.h"
#include "shell_encryption/rns/crt_interpolation.h"
#include "shell_encryption/rns/error_distribution.h"
#include "shell_encryption/rns/message_packing.h"
#include "shell_encryption/rns/rns_modulus.h"
#include "shell_encryption/sampler/discrete_gaussian.h"
#include "shell_wrapper/kahe.rs.h"
#include "shell_wrapper/shell_aliases.h"
#include "shell_wrapper/shell_types.h"
#include "shell_wrapper/shell_types.rs.h"
#include "shell_wrapper/single_thread_hkdf.rs.h"
#include "shell_wrapper/status.h"
#include "shell_wrapper/status.rs.h"
#include "shell_wrapper/status_macros.h"

using secure_aggregation::MakeFfiStatus;

namespace secure_aggregation {

absl::StatusOr<KahePublicParameters> CreateKahePublicParameters(
    const RnsContextConfig& rns_context_config, int log_kahe_plaintext_modulus,
    int num_public_polynomials, absl::string_view public_seed) {
  // Create RNS context (instantiated with BGV)
  SECAGG_ASSIGN_OR_RETURN(
      auto rns_context,
      RnsContext::Create(rns_context_config.log_n, rns_context_config.qs,
                         /*ps=*/{}, rns_context_config.t));
  auto rns_context_ptr =
      std::make_unique<const RnsContext>(std::move(rns_context));
  std::vector<const rlwe::PrimeModulus<ModularInt>*> moduli =
      rns_context_ptr->MainPrimeModuli();

  // Create RNS representation for plaintext modulus
  BigInteger kahe_plaintext_modulus = static_cast<BigInteger>(1)
                                      << log_kahe_plaintext_modulus;
  std::vector<ModularInt> zs;
  for (int i = 0; i < moduli.size(); ++i) {
    const auto mod_params_qi = moduli[i]->ModParams();
    Integer qi = mod_params_qi->modulus;
    BigInteger qi_big = static_cast<BigInteger>(qi);
    BigInteger plaintext_modulus_mod_qi = kahe_plaintext_modulus % qi_big;
    SECAGG_ASSIGN_OR_RETURN(
        ModularInt z,
        ModularInt::ImportInt(static_cast<Integer>(plaintext_modulus_mod_qi),
                              mod_params_qi));
    zs.push_back(std::move(z));
  }
  auto plaintext_modulus_rns = RnsInt{zs};

  // Generate public parameters ("a").
  auto prg_public_prng = Prng::Create(public_seed).value();
  std::vector<RnsPolynomial> public_polynomials;
  public_polynomials.reserve(num_public_polynomials);
  for (int i = 0; i < num_public_polynomials; ++i) {
    auto a = RnsPolynomial::SampleUniform(rns_context_config.log_n,
                                          prg_public_prng.get(), moduli)
                 .value();
    public_polynomials.push_back(std::move(a));
  }

  // Create an encoder and an error sampler that will be used in each call to
  // encrypt/decrypt.
  SECAGG_ASSIGN_OR_RETURN(
      auto prg_encoder,
      rlwe::CoefficientEncoder<ModularInt>::Create(rns_context_ptr.get()));
  SECAGG_ASSIGN_OR_RETURN(
      auto error_dg_sampler,
      rlwe::DiscreteGaussianSampler<Integer>::Create(kPrgErrorS));

  std::vector<BigInteger> modulus_hats =
      rlwe::RnsModulusComplements<ModularInt, BigInteger>(moduli);

  auto level = moduli.size() - 1;
  SECAGG_ASSIGN_OR_RETURN(std::vector<ModularInt> modulus_hats_invs,
                          rns_context_ptr->MainPrimeModulusCrtFactors(level));

  return KahePublicParameters{
      .context = std::move(rns_context_ptr),
      .plaintext_modulus = kahe_plaintext_modulus,
      .plaintext_modulus_rns = plaintext_modulus_rns,
      .moduli = std::move(moduli),
      .modulus_hats = std::move(modulus_hats),
      .modulus_hats_invs = std::move(modulus_hats_invs),
      .encoder = std::move(prg_encoder),
      .public_polynomials = std::move(public_polynomials),
      .error_dg_sampler = std::move(error_dg_sampler),
  };
}

absl::StatusOr<RnsPolynomial> GenerateSecretKey(
    const KahePublicParameters& params, Prng* prng) {
  // Create discrete Gaussian sampler.
  SECAGG_ASSIGN_OR_RETURN(
      auto secret_sampler,
      rlwe::DiscreteGaussianSampler<Integer>::Create(kPrgSecretS));

  // Sample secret coefficients, and encode as RNS polynomial.
  int num_prg_coeffs = 1 << params.context->LogN();
  std::vector<std::vector<ModularInt>> prg_secret_rns_coeffs(
      params.moduli.size());
  for (int j = 0; j < num_prg_coeffs; ++j) {
    SECAGG_ASSIGN_OR_RETURN(Integer value, secret_sampler->Sample(*prng));
    // Convert `value` to balanced representation mod q_i for each q_i.
    for (int i = 0; i < params.moduli.size(); ++i) {
      Integer value_balanced_qi = value;
      if (value > rlwe::DiscreteGaussianSampler<Integer>::kNegativeThreshold) {
        // `value` is a negative Gaussian sample, represented as a large
        // unsigned integer.
        value_balanced_qi = params.moduli[i]->Modulus() - (-value);
      }
      SECAGG_ASSIGN_OR_RETURN(
          ModularInt value_mod_qi,
          ModularInt::ImportInt(value_balanced_qi,
                                params.moduli[i]->ModParams()));
      prg_secret_rns_coeffs[i].push_back(value_mod_qi);
    }
  }
  SECAGG_ASSIGN_OR_RETURN(
      auto prg_secret, RnsPolynomial::Create(std::move(prg_secret_rns_coeffs),
                                             /*is_ntt=*/false));

  if (!prg_secret.IsNttForm()) {
    SECAGG_RETURN_IF_ERROR(prg_secret.ConvertToNttForm(params.moduli));
  }

  return prg_secret;
}

namespace internal {

absl::StatusOr<RnsPolynomial> EncryptPolynomial(
    const RnsPolynomial& plaintext, const RnsPolynomial& secret_key,
    RnsInt plaintext_modulus_rns, int log_n, const RnsPolynomial& a,
    absl::Span<const rlwe::PrimeModulus<ModularInt>* const> moduli,
    const rlwe::DiscreteGaussianSampler<Integer>* dg_sampler, Prng* prng) {
  // Sample the error term e (mod Q) from the error distribution.
  SECAGG_ASSIGN_OR_RETURN(RnsPolynomial c,
                          rlwe::SampleDiscreteGaussian<ModularInt>(
                              log_n, moduli, dg_sampler, prng));

  // c = e * t (mod Q).
  SECAGG_RETURN_IF_ERROR(c.MulInPlace(plaintext_modulus_rns, moduli));

  // c = e * t + m (mod Q).
  SECAGG_RETURN_IF_ERROR(c.AddInPlace(plaintext, moduli));
  if (!c.IsNttForm()) {
    SECAGG_RETURN_IF_ERROR(c.ConvertToNttForm(moduli));
  }

  // c = e * t + m + a * key (mod Q).
  SECAGG_RETURN_IF_ERROR(c.FusedMulAddInPlace(a, secret_key, moduli));
  return c;
}

absl::StatusOr<RnsPolynomial> DecryptPolynomial(
    const RnsPolynomial& ciphertext, const RnsPolynomial& secret_key,
    const RnsPolynomial& a,
    absl::Span<const rlwe::PrimeModulus<ModularInt>* const> moduli) {
  // p = a * k (mod Q).
  SECAGG_ASSIGN_OR_RETURN(RnsPolynomial p, secret_key.Mul(a, moduli));

  // p = - a * k (mod Q).
  SECAGG_RETURN_IF_ERROR(p.NegateInPlace(moduli));

  // p = c - a * k (mod Q).
  SECAGG_RETURN_IF_ERROR(p.AddInPlace(ciphertext, moduli));
  return p;
}

}  // namespace internal

absl::StatusOr<std::vector<RnsPolynomial>> EncodeAndEncryptVector(
    const std::vector<BigInteger>& packed_values,
    const RnsPolynomial& secret_key, const KahePublicParameters& params,
    Prng* prng) {
  std::vector<std::vector<BigInteger>> plaintexts;
  plaintexts.reserve(params.public_polynomials.size());
  int num_coeffs = 1 << params.context->LogN();

  for (size_t i = 0; i < packed_values.size(); i += num_coeffs) {
    size_t chunk_end = std::min<size_t>(packed_values.size(), i + num_coeffs);
    plaintexts.emplace_back(packed_values.begin() + i,
                            packed_values.begin() + chunk_end);
  }
  if (plaintexts.size() > params.public_polynomials.size()) {
    return absl::InvalidArgumentError("input too long.");
  }

  std::vector<RnsPolynomial> ciphertexts;
  for (int i = 0; i < plaintexts.size(); ++i) {
    const auto& packed_message = plaintexts[i];
    const RnsPolynomial& a = params.public_polynomials[i];
    // EncodeBgv will pad `packed_message` with zeros to the length of a
    // polynomial coefficient vector.
    SECAGG_ASSIGN_OR_RETURN(
        RnsPolynomial plaintext,
        params.encoder.EncodeBgv<BigInteger>(
            packed_message, params.plaintext_modulus, params.moduli));
    SECAGG_ASSIGN_OR_RETURN(
        RnsPolynomial ciphertext,
        internal::EncryptPolynomial(plaintext, secret_key,
                                    params.plaintext_modulus_rns,
                                    params.context->LogN(), a, params.moduli,
                                    params.error_dg_sampler.get(), prng));

    ciphertexts.push_back(std::move(ciphertext));
  }
  return ciphertexts;
}

absl::StatusOr<std::vector<BigInteger>> DecodeAndDecryptVector(
    absl::Span<const RnsPolynomial> ciphertexts,
    const RnsPolynomial& secret_key, const KahePublicParameters& params) {
  if (ciphertexts.size() > params.public_polynomials.size()) {
    return absl::InvalidArgumentError(
        "The size of `ciphertexts` cannot be larger than the size of public "
        "polynomials.");
  }
  std::vector<BigInteger> all_packed_messages;
  for (int i = 0; i < ciphertexts.size(); ++i) {
    const auto& ciphertext = ciphertexts[i];
    const RnsPolynomial& a = params.public_polynomials[i];
    SECAGG_ASSIGN_OR_RETURN(
        RnsPolynomial plaintext,
        internal::DecryptPolynomial(ciphertext, secret_key, a, params.moduli));
    SECAGG_ASSIGN_OR_RETURN(
        std::vector<BigInteger> packed_messages,
        params.encoder.DecodeBgv<BigInteger>(
            std::move(plaintext), params.plaintext_modulus, params.moduli,
            params.modulus_hats, params.modulus_hats_invs));
    all_packed_messages.insert(all_packed_messages.end(),
                               packed_messages.begin(), packed_messages.end());
  }
  return all_packed_messages;
}

}  // namespace secure_aggregation

FfiStatus CreateKahePublicParametersWrapper(uint64_t log_n, uint64_t log_t,
                                            rust::Slice<const uint64_t> qs,
                                            uint64_t num_public_polynomials,
                                            rust::Slice<const uint8_t> seed,
                                            KahePublicParametersWrapper* out) {
  if (out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  // Parse the RNS config
  constexpr int dummy_rns_plaintext_modulus = 2;  // Unused for KAHE
  secure_aggregation::RnsContextConfig rns_context_config =
      secure_aggregation::ParseRnsContextConfig(
          log_n, dummy_rns_plaintext_modulus, qs.data(), qs.size());

  auto statusor = secure_aggregation::CreateKahePublicParameters(
      rns_context_config, log_t, num_public_polynomials,
      absl::string_view(reinterpret_cast<const char*>(seed.data()),
                        seed.size()));
  if (!statusor.ok()) {
    return MakeFfiStatus(statusor.status());
  }
  out->ptr = std::make_unique<secure_aggregation::KahePublicParameters>(
      std::move(statusor.value()));
  return MakeFfiStatus();
}

FfiStatus GenerateSecretKeyWrapper(const KahePublicParametersWrapper& params,
                                   SingleThreadHkdfWrapper* prng,
                                   RnsPolynomialWrapper* out) {
  if (prng == nullptr || prng->ptr == nullptr || params.ptr == nullptr ||
      out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  auto statusor =
      secure_aggregation::GenerateSecretKey(*params.ptr, prng->ptr.get());
  if (!statusor.ok()) {
    return MakeFfiStatus(statusor.status());
  }
  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(statusor.value()));
  return MakeFfiStatus();
}

FfiStatus PackMessagesRaw(rust::Slice<const uint64_t> messages,
                          uint64_t packing_base, uint64_t packing_dimension,
                          uint64_t num_packed_values,
                          BigIntVectorWrapper* packed_values) {
  // Validate the wrappers.
  if (packed_values == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  // Allocate the vector for output packed values if needed.
  if (packed_values->ptr == nullptr) {
    packed_values->ptr =
        std::make_unique<std::vector<secure_aggregation::BigInteger>>();
  }
  auto curr_packed_values =
      rlwe::PackMessagesFlat<secure_aggregation::Integer,
                             secure_aggregation::BigInteger>(
          absl::MakeSpan(messages.data(), messages.size()), packing_base,
          packing_dimension);
  if (curr_packed_values.size() > num_packed_values) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        "The number of packed values exceeds `num_packed_values`."));
  }
  // Pad with zeros if needed.
  curr_packed_values.resize(num_packed_values, 0);
  // Append the packed values to the end of the output vector.
  packed_values->ptr->insert(packed_values->ptr->end(),
                             curr_packed_values.begin(),
                             curr_packed_values.end());
  return MakeFfiStatus();
}

FfiStatus UnpackMessagesRaw(uint64_t packing_base, uint64_t packing_dimension,
                            uint64_t num_packed_values,
                            BigIntVectorWrapper& packed_values,
                            rust::Vec<uint64_t>& out) {
  // Validate the wrappers.
  if (packed_values.ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  if (packed_values.ptr->size() < num_packed_values) {
    return MakeFfiStatus(
        absl::InvalidArgumentError("insufficient number of packed values."));
  }
  std::vector<uint64_t> unpacked_messages =
      rlwe::UnpackMessagesFlat<secure_aggregation::Integer,
                               secure_aggregation::BigInteger>(
          absl::MakeSpan(*packed_values.ptr).subspan(0, num_packed_values),
          packing_base, packing_dimension);
  packed_values.ptr->erase(packed_values.ptr->begin(),
                           packed_values.ptr->begin() + num_packed_values);
  for (auto& val : unpacked_messages) {
    out.push_back(val);
  }
  return MakeFfiStatus();
}

FfiStatus Encrypt(const BigIntVectorWrapper& packed_values,
                  const RnsPolynomialWrapper& secret_key,
                  const KahePublicParametersWrapper& params,
                  SingleThreadHkdfWrapper* prng, RnsPolynomialVecWrapper* out) {
  // Validate the wrappers.
  if (packed_values.ptr == nullptr || secret_key.ptr == nullptr ||
      params.ptr == nullptr || prng == nullptr || prng->ptr == nullptr ||
      out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  auto ciphertext_vec = secure_aggregation::EncodeAndEncryptVector(
      *packed_values.ptr, *secret_key.ptr, *params.ptr, prng->ptr.get());

  if (!ciphertext_vec.ok()) {
    return MakeFfiStatus(ciphertext_vec.status());
  }
  out->len = ciphertext_vec.value().size();
  out->ptr = std::make_unique<std::vector<secure_aggregation::RnsPolynomial>>(
      std::move(ciphertext_vec.value()));
  return MakeFfiStatus();
}

FfiStatus Decrypt(const RnsPolynomialVecWrapper& ciphertexts,
                  const RnsPolynomialWrapper& secret_key,
                  const KahePublicParametersWrapper& params,
                  BigIntVectorWrapper* output_values) {
  // Validate the wrappers.
  if (secret_key.ptr == nullptr || params.ptr == nullptr ||
      ciphertexts.ptr == nullptr || output_values == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  // Secret key may be in coeff form after deserialization from an AHE
  // plaintext, so convert to NTT form if needed.
  if (!secret_key.ptr->IsNttForm()) {
    auto status = secret_key.ptr->ConvertToNttForm(params.ptr->moduli);
    if (!status.ok()) {
      return MakeFfiStatus(status);
    }
  }

  auto decrypted_values = secure_aggregation::DecodeAndDecryptVector(
      *ciphertexts.ptr, *secret_key.ptr, *params.ptr);
  if (!decrypted_values.ok()) {
    return MakeFfiStatus(decrypted_values.status());
  }
  output_values->ptr =
      std::make_unique<std::vector<secure_aggregation::BigInteger>>(
          std::move(decrypted_values.value()));
  return MakeFfiStatus();
}
